lib/src/protocol.dart

```diff
@@ -0,0 +1,270 @@
+      wroteData = true;
+    }
+
+    if (_childElementUpdates.isNotEmpty) {
+      final jsUpdates = <Map<String, dynamic>>[];
+      for (ElementUpdate update in _childElementUpdates) {
+        final childUpdate = <String, dynamic>{};
+        if (update.render(childUpdate)) {
+          jsUpdates.add(childUpdate);
+        }
+      }
+
+      if (jsUpdates.isNotEmpty) {
+        js["update-elements"] = jsUpdates;
+        wroteData = true;
+      }
+    }
+
+    if (_attributes.isNotEmpty) {
+      final jsAttrUpdates = <String, dynamic>{};
+      for (AttributeUpdate attrUpdate in _attributes) {
+        jsAttrUpdates[attrUpdate.name] = attrUpdate.value;
+      }
+      js["attrs"] = jsAttrUpdates;
+      wroteData = true;
+    }
+
+    if (_classNames.isNotEmpty) {
+      final jsClassNames = <String>[];
+      for (String className in _classNames) {
+        jsClassNames.add(className);
+      }
+      js["classes"] = jsClassNames;
+      wroteData = true;
+    }
+
+    if (wroteData) {
+      js["index"] = _index;
+    }
+
+    return wroteData;
+  }
+
+  /// Assumes that this element update is exlusively made of insertions and
+  /// renders it as a plain HTML into the given [buffer].
+  void printHtml(StringBuffer buf) {
+    if (_index != -1) {  // we don't print host tag.
+      buf..write("<")..write(_tag);
+
+      if (_key != "") {
+        buf..write(" _bkey=\"")..write(_key)..write("\"");
+      }
+
+      if (_attributes.isNotEmpty) {
+        for (final AttributeUpdate attr in _attributes) {
+          buf..write(" ")..write(attr.name)..write("=\"")..write(attr.value)..write("\"");
+        }
+      }
+
+      if (_classNames.isNotEmpty) {
+        buf..write(" class=\"");
+        for (int i = 0; i < _classNames.length; i++) {
+          buf..write(_classNames[i]);
+          if (i + 1 < _classNames.length) {
+            buf..write(" ");
+          }
+        }
+        buf..write("\"");
+      }
+
+      if (_bid != "") {
+        buf..write(" _bid=\"")..write(_bid)..write("\"");
+      }
+
+      buf..write(">");
+    }
+
+    if (_text != null) {
+      buf..write(_text);
+    }
+
+    for(final childElement in _childElementInsertions) {
+      childElement.printHtml(buf);
+    }
+
+    if (_index != -1) {
+      buf..write("</")..write(_tag)..write(">");
+    }
+  }
+
+  void removeChild(int index) { _removes.add(index); }
+
+  void moveChild(int insertionIndex, int moveFrom) {
+    _moves.add(new Move(insertionIndex, moveFrom));
+  }
+
+  ElementUpdate insertChildElement(int insertionIndex) {
+    _childElementInsertions.add(new ElementUpdate(insertionIndex));
+    return _childElementInsertions.last;
+  }
+
+  ElementUpdate updateChildElement(int index) {
+    _childElementUpdates.add(new ElementUpdate(index));
+    return _childElementUpdates.last;
+  }
+
+  void setTag(String tag) { _tag = tag; }
+
+  void setKey(Key key) { _key = key.toString(); }
+
+  void updateText(String text) { _text = text; }
```

Comment:
Ok, I now remember why I decided to differentiate between "set" and "update". Something you "set" never changes after it's been set for the very first time, such as tag name and key. This you "update" can change over time. But I don't think that a naming convention will really be that effective. Perhaps we should use assertions to make sure this doesn't actually happen.